### PR TITLE
atomspace memory management bugs

### PR DESCRIPTION
**Pull request by linas**
Revealed by the pattern-miner. per discussions at https://github.com/opencog/atomspace/pull/2027#issuecomment-463605231

### CODE REVIEW

**Comment by linas** [comment]
Hmm. After this change, CythonBackwardchainer is segfaulting in circleci, but it passes fine for me, so I don't know how to debug. 

One the one hand, ts plausible that this causes a segfault (atomsaces that appear to be unused get deleted) on the other hand, this is much more conservative about when it allows deletes to happen...

**Comment by linas** [comment]
Hmm. CythonBackwardChainer has a mem leak. It's blocking the other pull req #2074 also

File: opencog/guile/SchemeSmobAS.cc [hunk 1]
```diff
@@ -75,14 +75,16 @@ void SchemeSmob::release_as (AtomSpace *as)
...
 		// (Recursively) decrement the use count on the parent.
 		// We had incremented it earlier, in `ss_new_as`, when
-		// creating it.
+		// creating it. Do not delete it, unless the guile gc
+		// asks us to.
 		if (env and deleteable_as.end() != deleteable_as.find(env))
 		{
-			--deleteable_as[env];
-			if (0 == deleteable_as[env]) release_as(env);
+			if (0 < deleteable_as[env])
```
Comment (vsbogd) [inline opencog/guile/SchemeSmobAS.cc]: Do we really have the situations when `deleteable_as[env] == 0` here? Or it is double check guard? Should we log such situations for analysis?

**Comment by vsbogd** [comment]
For segfault is reproducible in Release version.

**Comment by vsbogd** [comment]
The stack trace is:
<details>
<pre>
#0  0x00007ffff5e191a6 in opencog::AtomSpace::get_environ (this=0xc97ba0)
    at /home/opencog/projects/atomspace/opencog/atomspace/AtomSpace.h:108
#1  opencog::SchemeSmob::as_ref_count (old_as=<optimized out>, nas=nas@entry=0xd51550)
    at /home/opencog/projects/atomspace/opencog/guile/SchemeSmobAS.cc:329
#2  0x00007ffff5e19474 in opencog::SchemeSmob::ss_set_env_as (nas=0xd51550)
    at /home/opencog/projects/atomspace/opencog/guile/SchemeSmobAS.cc:368
#3  0x00007ffff5e0d979 in opencog::SchemeEval::do_eval (this=0xc9c110, 
    expr="(add-to-load-path \"../../..\")")
    at /home/opencog/projects/atomspace/opencog/guile/SchemeEval.cc:561
#4  0x00007ffff5e0db2a in opencog::SchemeEval::c_wrap_eval (p=0xc9c110)
    at /home/opencog/projects/atomspace/opencog/guile/SchemeEval.cc:507
#5  0x00007ffff58cf68a in c_body (d=0x7fffffffc270) at ../../libguile/continuations.c:422
#6  0x00007ffff595263b in vm_regular_engine (thread=0x1, vp=0x7fffea790ea0, registers=0x0, 
    resume=13204384) at ../../libguile/vm-engine.c:784
#7  0x00007ffff59556b2 in scm_call_n (proc=proc@entry=0x7fffe89bb7c0, argv=argv@entry=0x0, 
    nargs=nargs@entry=0) at ../../libguile/vm.c:1257
#8  0x00007ffff58d5619 in scm_call_0 (proc=proc@entry=0x7fffe89bb7c0)
    at ../../libguile/eval.c:481
#9  0x00007ffff59445a9 in catch (tag=tag@entry=0x404, thunk=0x7fffe89bb7c0, 
    handler=0x7fffe89bb7a0, pre_unwind_handler=0x7fffe89bb780) at ../../libguile/throw.c:137
#10 0x00007ffff5944915 in scm_catch_with_pre_unwind_handler (key=key@entry=0x404, 
    thunk=<optimized out>, handler=<optimized out>, pre_unwind_handler=<optimized out>)
    at ../../libguile/throw.c:254
#11 0x00007ffff5944acf in scm_c_catch (tag=tag@entry=0x404, 
    body=body@entry=0x7ffff58cf680 <c_body>, body_data=body_data@entry=0x7fffffffc270, 
    handler=handler@entry=0x7ffff58cf920 <c_handler>, 
    handler_data=handler_data@entry=0x7fffffffc270, 
    pre_unwind_handler=pre_unwind_handler@entry=0x7ffff58cf780 <pre_unwind_handler>, 
    pre_unwind_handler_data=0x7fffea78bb60) at ../../libguile/throw.c:377
#12 0x00007ffff58cfc90 in scm_i_with_continuation_barrier (
    body=body@entry=0x7ffff58cf680 <c_body>, body_data=body_data@entry=0x7fffffffc270, 
    handler=handler@entry=0x7ffff58cf920 <c_handler>, 
    handler_data=handler_data@entry=0x7fffffffc270, 
    pre_unwind_handler=pre_unwind_handler@entry=0x7ffff58cf780 <pre_unwind_handler>, 
    pre_unwind_handler_data=0x7fffea78bb60) at ../../libguile/continuations.c:360
#13 0x00007ffff58cfd75 in scm_c_with_continuation_barrier (func=<optimized out>, 
    data=<optimized out>) at ../../libguile/continuations.c:456
#14 0x00007ffff320c59f in GC_call_with_gc_active (
    fn=fn@entry=0x7ffff59426b0 <with_guile_trampoline>, 
    client_data=client_data@entry=0x7fffffffc350) at extra/../pthread_support.c:1433
#15 0x00007ffff59430c1 in with_guile (base=base@entry=0x7fffffffc320, 
    data=data@entry=0x7fffffffc350) at ../../libguile/threads.c:689
#16 0x00007ffff3205f12 in GC_call_with_stack_base (fn=fn@entry=0x7ffff5943030 <with_guile>, 
    arg=arg@entry=0x7fffffffc350) at extra/../misc.c:2132
#17 0x00007ffff5943468 in scm_i_with_guile (dynamic_state=<optimized out>, data=0xc9c110, 
    func=0x7ffff5e0db10 <opencog::SchemeEval::c_wrap_eval(void*)>)
    at ../../libguile/threads.c:704
#18 scm_with_guile (
    func=func@entry=0x7ffff5e0db10 <opencog::SchemeEval::c_wrap_eval(void*)>, 
    data=data@entry=0xc9c110) at ../../libguile/threads.c:710
#19 0x00007ffff5e0daee in opencog::SchemeEval::eval_expr (this=0xc9c110, expr=...)
    at /home/opencog/projects/atomspace/opencog/guile/SchemeEval.cc:479
#20 0x00007ffff66c26a6 in opencog::SchemeEval::eval (expr="(add-to-load-path \"../../..\")", 
    this=0xc9c110) at /home/opencog/projects/atomspace/opencog/guile/SchemeEval.h:161
#21 opencog::eval_scheme (as=<optimized out>, s="(add-to-load-path \"../../..\")")
    at /home/opencog/projects/atomspace/opencog/cython/opencog/PyScheme.cc:58
#22 0x00007ffff66c40ac in __pyx_pf_7opencog_14scheme_wrapper_scheme_eval (
    __pyx_self=<optimized out>, __pyx_v_a=<optimized out>, __pyx_v_pys=<optimized out>)
    at /home/opencog/projects/atomspace-build/opencog/cython/opencog/scheme_wrapper.cpp:1063
#23 __pyx_pw_7opencog_14scheme_wrapper_1scheme_eval (__pyx_self=<optimized out>, 
    __pyx_args=<optimized out>, __pyx_kwds=<optimized out>)
    at /home/opencog/projects/atomspace-build/opencog/cython/opencog/scheme_wrapper.cpp:1015
#24 0x00000000004ea137 in PyCFunction_Call () at ../Objects/methodobject.c:98
#25 0x0000000000536d94 in call_function (oparg=<optimized out>, pp_stack=0x7fffffffc5d0)
    at ../Python/ceval.c:4705
#26 PyEval_EvalFrameEx () at ../Python/ceval.c:3236
#27 0x000000000053b294 in fast_function (nk=<optimized out>, na=<optimized out>, 
    n=<optimized out>, pp_stack=0x7fffffffc700, func=<optimized out>)
    at ../Python/ceval.c:4803
#28 call_function (oparg=<optimized out>, pp_stack=0x7fffffffc700) at ../Python/ceval.c:4730
#29 PyEval_EvalFrameEx () at ../Python/ceval.c:3236
#30 0x000000000053b294 in fast_function (nk=<optimized out>, na=<optimized out>, 
    n=<optimized out>, pp_stack=0x7fffffffc830, func=<optimized out>)
    at ../Python/ceval.c:4803
#31 call_function (oparg=<optimized out>, pp_stack=0x7fffffffc830) at ../Python/ceval.c:4730
#32 PyEval_EvalFrameEx () at ../Python/ceval.c:3236
#33 0x0000000000540b0b in _PyEval_EvalCodeWithName (qualname=0x0, name=0x0, closure=0x0, 
    kwdefs=0x0, defcount=1, defs=0x7ffff69a1f08, kwcount=0, kws=<optimized out>, 
    argcount=<optimized out>, args=<optimized out>, locals=<optimized out>, 
    globals=<optimized out>, _co=<code at remote 0x7ffff6a838a0>) at ../Python/ceval.c:4018
#34 PyEval_EvalCodeEx () at ../Python/ceval.c:4039
#35 0x00000000004ec3f7 in function_call.lto_priv () at ../Objects/funcobject.c:627
#36 0x00000000005c20e7 in PyObject_Call () at ../Objects/abstract.c:2165
#37 0x0000000000538cab in ext_do_call (nk=<optimized out>, na=1, flags=<optimized out>, 
    pp_stack=0x7fffffffcae8, func=<function at remote 0x7ffff69ab620>)
    at ../Python/ceval.c:5034
#38 PyEval_EvalFrameEx () at ../Python/ceval.c:3275
#39 0x0000000000540b0b in _PyEval_EvalCodeWithName (qualname=0x0, name=0x0, closure=0x0, 
    kwdefs=0x0, defcount=0, defs=0x0, kwcount=0, kws=<optimized out>, 
    argcount=<optimized out>, args=<optimized out>, locals=<optimized out>, 
    globals=<optimized out>, _co=<code at remote 0x7ffff6a839c0>) at ../Python/ceval.c:4018
#40 PyEval_EvalCodeEx () at ../Python/ceval.c:4039
#41 0x00000000004ec2e3 in function_call.lto_priv () at ../Objects/funcobject.c:627
#42 0x00000000005c20e7 in PyObject_Call () at ../Objects/abstract.c:2165
#43 0x00000000004fbfce in method_call.lto_priv () at ../Objects/classobject.c:330
#44 0x00000000005c20e7 in PyObject_Call () at ../Objects/abstract.c:2165
#45 0x0000000000574db6 in slot_tp_call () at ../Objects/typeobject.c:6053
#46 0x00000000005c20e7 in PyObject_Call () at ../Objects/abstract.c:2165
#47 0x000000000053b656 in do_call (nk=<optimized out>, na=<optimized out>, 
    pp_stack=0x7fffffffce80, func=<optimized out>) at ../Python/ceval.c:4936
#48 call_function (oparg=<optimized out>, pp_stack=0x7fffffffce80) at ../Python/ceval.c:4732
#49 PyEval_EvalFrameEx () at ../Python/ceval.c:3236
#50 0x0000000000540b0b in _PyEval_EvalCodeWithName (qualname=0x0, name=0x0, closure=0x0, 
    kwdefs=0x0, defcount=1, defs=0x7ffff69b4060, kwcount=0, kws=<optimized out>, 
    argcount=<optimized out>, args=<optimized out>, locals=<optimized out>, 
    globals=<optimized out>, _co=<code at remote 0x7ffff69a0ae0>) at ../Python/ceval.c:4018
#51 PyEval_EvalCodeEx () at ../Python/ceval.c:4039
#52 0x00000000004ec3f7 in function_call.lto_priv () at ../Objects/funcobject.c:627
#53 0x00000000005c20e7 in PyObject_Call () at ../Objects/abstract.c:2165
#54 0x0000000000538cab in ext_do_call (nk=<optimized out>, na=1, flags=<optimized out>, 
    pp_stack=0x7fffffffd138, func=<function at remote 0x7ffff69b3378>)
    at ../Python/ceval.c:5034
#55 PyEval_EvalFrameEx () at ../Python/ceval.c:3275
#56 0x0000000000540b0b in _PyEval_EvalCodeWithName (qualname=0x0, name=0x0, closure=0x0, 
    kwdefs=0x0, defcount=0, defs=0x0, kwcount=0, kws=<optimized out>, 
    argcount=<optimized out>, args=<optimized out>, locals=<optimized out>, 
    globals=<optimized out>, _co=<code at remote 0x7ffff6a92030>) at ../Python/ceval.c:4018
#57 PyEval_EvalCodeEx () at ../Python/ceval.c:4039
#58 0x00000000004ec2e3 in function_call.lto_priv () at ../Objects/funcobject.c:627
#59 0x00000000005c20e7 in PyObject_Call () at ../Objects/abstract.c:2165
#60 0x00000000004fbfce in method_call.lto_priv () at ../Objects/classobject.c:330
#61 0x00000000005c20e7 in PyObject_Call () at ../Objects/abstract.c:2165
#62 0x0000000000574db6 in slot_tp_call () at ../Objects/typeobject.c:6053
#63 0x00000000005c20e7 in PyObject_Call () at ../Objects/abstract.c:2165
#64 0x000000000053b656 in do_call (nk=<optimized out>, na=<optimized out>, 
    pp_stack=0x7fffffffd4d0, func=<optimized out>) at ../Python/ceval.c:4936
#65 call_function (oparg=<optimized out>, pp_stack=0x7fffffffd4d0) at ../Python/ceval.c:4732
#66 PyEval_EvalFrameEx () at ../Python/ceval.c:3236
#67 0x0000000000540b0b in _PyEval_EvalCodeWithName (qualname=0x0, name=0x0, closure=0x0, 
    kwdefs=0x0, defcount=1, defs=0x7ffff69b4060, kwcount=0, kws=<optimized out>, 
    argcount=<optimized out>, args=<optimized out>, locals=<optimized out>, 
    globals=<optimized out>, _co=<code at remote 0x7ffff69a0ae0>) at ../Python/ceval.c:4018
#68 PyEval_EvalCodeEx () at ../Python/ceval.c:4039
#69 0x00000000004ec3f7 in function_call.lto_priv () at ../Objects/funcobject.c:627
#70 0x00000000005c20e7 in PyObject_Call () at ../Objects/abstract.c:2165
#71 0x0000000000538cab in ext_do_call (nk=<optimized out>, na=1, flags=<optimized out>, 
    pp_stack=0x7fffffffd788, func=<function at remote 0x7ffff69b3378>)
    at ../Python/ceval.c:5034
#72 PyEval_EvalFrameEx () at ../Python/ceval.c:3275
#73 0x0000000000540b0b in _PyEval_EvalCodeWithName (qualname=0x0, name=0x0, closure=0x0, 
    kwdefs=0x0, defcount=0, defs=0x0, kwcount=0, kws=<optimized out>, 
    argcount=<optimized out>, args=<optimized out>, locals=<optimized out>, 
    globals=<optimized out>, _co=<code at remote 0x7ffff6a92030>) at ../Python/ceval.c:4018
#74 PyEval_EvalCodeEx () at ../Python/ceval.c:4039
#75 0x00000000004ec2e3 in function_call.lto_priv () at ../Objects/funcobject.c:627
#76 0x00000000005c20e7 in PyObject_Call () at ../Objects/abstract.c:2165
#77 0x00000000004fbfce in method_call.lto_priv () at ../Objects/classobject.c:330
#78 0x00000000005c20e7 in PyObject_Call () at ../Objects/abstract.c:2165
#79 0x0000000000574db6 in slot_tp_call () at ../Objects/typeobject.c:6053
#80 0x00000000005c20e7 in PyObject_Call () at ../Objects/abstract.c:2165
#81 0x000000000053b656 in do_call (nk=<optimized out>, na=<optimized out>, 
    pp_stack=0x7fffffffdb20, func=<optimized out>) at ../Python/ceval.c:4936
#82 call_function (oparg=<optimized out>, pp_stack=0x7fffffffdb20) at ../Python/ceval.c:4732
#83 PyEval_EvalFrameEx () at ../Python/ceval.c:3236
#84 0x000000000053b294 in fast_function (nk=<optimized out>, na=<optimized out>, 
    n=<optimized out>, pp_stack=0x7fffffffdc50, func=<optimized out>)
    at ../Python/ceval.c:4803
#85 call_function (oparg=<optimized out>, pp_stack=0x7fffffffdc50) at ../Python/ceval.c:4730
#86 PyEval_EvalFrameEx () at ../Python/ceval.c:3236
#87 0x000000000053b294 in fast_function (nk=<optimized out>, na=<optimized out>, 
    n=<optimized out>, pp_stack=0x7fffffffdd80, func=<optimized out>)
    at ../Python/ceval.c:4803
#88 call_function (oparg=<optimized out>, pp_stack=0x7fffffffdd80) at ../Python/ceval.c:4730
#89 PyEval_EvalFrameEx () at ../Python/ceval.c:3236
#90 0x000000000054124a in _PyEval_EvalCodeWithName (qualname=0x0, name=0x0, closure=0x0, 
    kwdefs={'tb_locals': False}, defcount=11, defs=0x7ffff6937170, kwcount=0, 
    kws=<optimized out>, argcount=<optimized out>, args=<optimized out>, 
    locals=<optimized out>, globals=<optimized out>, _co=<code at remote 0x7ffff69badb0>)
    at ../Python/ceval.c:4018
#91 PyEval_EvalCodeEx () at ../Python/ceval.c:4039
#92 0x00000000004ec358 in function_call.lto_priv () at ../Objects/funcobject.c:627
#93 0x00000000005c20e7 in PyObject_Call () at ../Objects/abstract.c:2165
#94 0x00000000004fbfce in method_call.lto_priv () at ../Objects/classobject.c:330
#95 0x00000000005c20e7 in PyObject_Call () at ../Objects/abstract.c:2165
#96 0x0000000000574c19 in slot_tp_init () at ../Objects/typeobject.c:6268
#97 0x000000000057f58c in type_call.lto_priv () at ../Objects/typeobject.c:905
#98 0x00000000005c20e7 in PyObject_Call () at ../Objects/abstract.c:2165
#99 0x000000000053b656 in do_call (nk=<optimized out>, na=<optimized out>, 
    pp_stack=0x7fffffffe150, func=<optimized out>) at ../Python/ceval.c:4936
#100 call_function (oparg=<optimized out>, pp_stack=0x7fffffffe150) at ../Python/ceval.c:4732
#101 PyEval_EvalFrameEx () at ../Python/ceval.c:3236
#102 0x000000000053fc97 in _PyEval_EvalCodeWithName () at ../Python/ceval.c:4018
#103 0x00000000005409bf in PyEval_EvalCodeEx () at ../Python/ceval.c:4039
#104 PyEval_EvalCode (co=<optimized out>, globals=<optimized out>, locals=<optimized out>)
    at ../Python/ceval.c:777
#105 0x000000000060cb42 in run_mod () at ../Python/pythonrun.c:976
#106 0x000000000060efea in PyRun_FileExFlags () at ../Python/pythonrun.c:929
#107 0x000000000060f7dc in PyRun_SimpleFileExFlags () at ../Python/pythonrun.c:396
#108 0x0000000000640256 in run_file (p_cf=0x7fffffffe3c0, 
    filename=0xa74280 L"../atomspace/tests/cython/backwardchainer/test_bc.py", fp=0xb2f0f0)
    at ../Modules/main.c:318
#109 Py_Main () at ../Modules/main.c:768
#110 0x00000000004d0001 in main () at ../Programs/python.c:65
#111 0x00007ffff7810830 in __libc_start_main (main=0x4cff20 <main>, argc=2, 
    argv=0x7fffffffe5d8, init=<optimized out>, fini=<optimized out>, 
    rtld_fini=<optimized out>, stack_end=0x7fffffffe5c8) at ../csu/libc-start.c:291
#112 0x00000000005d6999 in _start ()
</pre>
</details>

**Comment by vsbogd** [comment]
ok, the reason of the segfault is that when atomspace is released by Sheme GC at the end of the first python test its pointer stays associated with `atomspace_fuild`. When next python test starts it tries to call `ss_set_env_as` with new atomspace and this call leads to the call of `get_environ` method on the deleted pointer at `SchemeSmobAS.cc:329`. Which in turn lead to segfault.

**Comment by vsbogd** [comment]
Memory leak in Python is probably caused by absence of `tearDown` method in `test_bc.py`, so Python atomspace stays initialized after test.

**Comment by vsbogd** [comment]
Raised https://github.com/opencog/atomspace/pull/2076 to finalize atomspace after test

File: opencog/guile/SchemeSmobAS.cc (hunk 1 above)
Comment (linas) [reply opencog/guile/SchemeSmobAS.cc]: A count of zero means that there are no known C++ users of the thing. There might still be guile references to it, that we don't know of.  The memory can only be collected if there are no known C++ users, and if guile thinks it doesn't have any users.

**Comment by linas** [comment]
I will look. guile should "know" about what's in a fluid, and therefore, should not be asking for it to be deleted.

**Comment by vsbogd** [comment]
I have checked stacktraces for `delete` and segfault.

**Comment by linas** [comment]
It feels like a guile bug now .. if there is some ref in some fluid, the GC should not be trying to collect that memory ... so that's weird/wrong, somehow. Can't quite figure out why.

**Comment by linas** [comment]
pull req #2074 exposed half-a-dozen variant use-after-free bugs; after fixing this, it seems that this now passes also.

**Comment by vsbogd** [comment]
`test_bc.py` still sporadically fails with segfault on my machine with the same stack trace

**Comment by vsbogd** [comment]
More frequent when run manually via `python3 ../../../tests/cython/backwardchainer/test_bc.py`

**Comment by ngeiswei** [comment]
BTW, guys, if you know an easy way to capture that bug in a unit test, I think that would be super good, because I suspect we already went through a cycle of fixing-then-breaking once.

**Comment by linas** [comment]
This seems to have something to do with circle-ci and probably has nothing to do with the backward chainer. On my machine I ran this unit test 100 times in a loop, it never failed once. I'm debugging a similar mystery crash in #2087

**Comment by vsbogd** [comment]
This [commit](https://github.com/opencog/atomspace/commit/78b9e1f298a4d0111f98db66d55c828e3c3d0b91) contains additional console messages and test which can reproduce incorrect behaviour. It doesn't crash but it can be seen that atomspace is used after deletion.
Command to run:
```
python3 ../atomspace/tests/cython/backwardchainer/test_crash.py
```
Output:
```
AtomSpace(): 0x2236f70 <-- Python atomspace
AtomSpace(): 0x225c540 <-- SchemeEval atomspace
ss_set_as: 0x225c540
as_ref_count oas: (nil)
as_ref_count nas: 0x225c540
ss_set_env_as: 0x2236f70
as_ref_count oas: 0x225c540
as_ref_count nas: 0x2236f70
nas->get_environ(): 0x2236f70
ss_set_env_as: 0x225c540
as_ref_count oas: 0x2236f70
as_ref_count nas: 0x225c540
oas->get_environ(): 0x2236f70
ss_set_env_as: 0x2236f70
as_ref_count oas: 0x225c540
as_ref_count nas: 0x2236f70
nas->get_environ(): 0x2236f70
~AtomSpace(): 0x225c540 <-- SchemeEval atomspace is deleted after (gc) call
ss_set_env_as: 0x225c540
as_ref_count oas: 0x2236f70
as_ref_count nas: 0x225c540
nas->get_environ(): 0x225c540 <-- method is called on object which is already deleted
oas->get_environ(): 0x2236f70
ss_set_env_as: 0x2236f70
as_ref_count oas: 0x225c540
as_ref_count nas: 0x2236f70
nas->get_environ(): 0x2236f70
oas->get_environ(): 0x225c540  <-- method is called on object which is already deleted
ss_set_env_as: 0x225c540
as_ref_count oas: 0x2236f70
as_ref_count nas: 0x225c540
nas->get_environ(): 0x225c540  <-- method is called on object which is already deleted
oas->get_environ(): 0x2236f70
~AtomSpace(): 0x2236f70
```
I cannot find a way to make it crash each time, because until memory block is not reused after deletion it will not crash.

**Comment by linas** [comment]
The above trace is almost surely the same one as in the cython bindiings for the backward chainer. The problem is here, near line 674 of `rule-engine/backwardchainer/BIT.cc`
```
BIT::BIT(AtomSpace& as,
         const Handle& target,
         const Handle& vardecl,
         const BITNodeFitness& fitness)
   : bit_as(&as), // child atomspace of as
     _as(&as),
```

So: the original atomspace `&as` gets deleted, but the backward chainer is holding on to a reference to a child atomspace that references this now-deleted atomspace.  One possible solution is to delete the backward-chainer *before* deleting the atomspace that it depends on.  There are other solutions, too.

**Comment by linas** [comment]
To be clear: I tried to fix this by calling the destructor for the backward chainer in the cython bindings. I did not try to go through the code-base to find every situation where the backward-chainer is being used -- there might be other use-after free scenarios popping up.  The primary issue is that the backward chainer is holding on to an object, without having any way of verifying whether that object exists or not.Thus, an alternative solution might be to delete the instance of `class BIT` the moment that chaining is done, before returning to the caller. @ngeiswei  would have a better idea of what it would take to implement that.

**Comment by linas** [comment]
OK, so I found a way to reliably reproduce this; am debugging now. The trick was to add `scheme_eval(self.atomspace, '(gc)')` to lots of places in the `test_bc.py` file

**Comment by linas** [comment]
Found it. Pull req #2091 fixes it.

**Pull request merged**